src/librustc_mir/transform/inline.rs

```diff
@@ -0,0 +1,880 @@
+
+        loop {
+            local_change = false;
+            let mut csi = 0;
+            while csi < callsites.len() {
+                let foreign_mir;
+
+                let callsite = callsites[csi];
+                csi += 1;
+
+                let callee_mir = {
+                    let callee_mir : Option<&Mir<'tcx>> = if callsite.callee.is_local() {
+                        map.map.get(&callsite.callee)
+                    } else {
+                        foreign_mir = self.get_foreign_mir(callsite.callee);
+                        foreign_mir.as_ref().map(|m| &**m)
+                    };
+
+                    let callee_mir = if let Some(m) = callee_mir {
+                        m
+                    } else {
+                        continue;
+                    };
+
+                    if !self.should_inline(callsite, callee_mir) {
+                        continue;
+                    }
+
+                    callee_mir.subst(self.tcx, callsite.substs)
+                };
+
+                let caller_mir = map.map.get_mut(&callsite.caller).unwrap();
+
+                let start = caller_mir.basic_blocks().len();
+
+                if !self.inline_call(callsite, caller_mir, callee_mir) {
+                    continue;
+                }
+
+                inlined_into.insert(callsite.caller);
+
+                // Add callsites from inlined function
+                for (bb, bb_data) in caller_mir.basic_blocks().iter_enumerated().skip(start) {
+                    // Only consider direct calls to functions
+                    let terminator = bb_data.terminator();
+                    if let TerminatorKind::Call {
+                        func: Operand::Constant(ref f), .. } = terminator.kind {
+                        if let ty::TyFnDef(callee_def_id, substs, _) = f.ty.sty {
+                            // Don't inline the same function multiple times.
+                            if callsite.callee != callee_def_id {
+                                callsites.push(CallSite {
+                                    caller: callsite.caller,
+                                    callee: callee_def_id,
+                                    substs: substs,
+                                    bb: bb,
+                                    location: terminator.source_info
+                                });
+                            }
+                        }
+                    }
+                }
+
+
+                csi -= 1;
+                if scc.len() == 1 {
+                    callsites.swap_remove(csi);
+                } else {
+                    callsites.remove(csi);
+                }
+
+                local_change = true;
+                changed = true;
+            }
+
+            if !local_change {
+                break;
+            }
+        }
+
+        // Simplify functions we inlined into.
+        for def_id in inlined_into {
+            let caller_mir = map.map.get_mut(&def_id).unwrap();
```

Comment:
I would prefer a more organized set of "post-inlining function passes", and to run it in the inner loop, rather than the outer one.